### PR TITLE
Cache encoded preference values.

### PR DESCRIPTION
**Pull request by phulin**
Each call to set a property results in writing all of the preferences out to disk, if `saveSettingsOnSet` is enabled. To my knowledge, this is the default and most users keep it enabled. This can take up a substantial amount of CPU runtime (say 10-15% in my unrepresentative profiling tests), as the properties need to all be encoded using `encodeString` each update.

This PR caches the encoded lines for each property, if `saveSettingsOnSet` is enabled. This means that only the changed property has to be re-encoded, rather than every single property. Saving the preferences to file then becomes a simple concatenation of the cached encoded lines. Because the encoded values are stored in a TreeMap, we can also get rid of the sorting I added when moving the main map to a ConcurrentHashMap.

The following microbenchmark runs in 78 seconds without this PR and 8.8 seconds with it - a roughly 9x speedup. I did not observe any speed change when `saveSettingsOnSet` was off.
```java
  @Test
  public void speed() {
    Preferences.setBoolean("saveSettingsOnSet", true);
    Preferences.saveSettingsToFile = true;
    long start = 0;
    ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
    var keySet = new ArrayList<>(Preferences.getMap(false, true).keySet());
    long threadId = Thread.currentThread().getId();
    for (int i = -2; i < 8; i++) {
      if (i == 0) start = threadMXBean.getThreadCpuTime(threadId);
      for (var pref : keySet) {
        Preferences.setString(pref, "abc" + i);
      }
    }
    long end = threadMXBean.getThreadCpuTime(threadId);
    System.out.println("Elapsed: " + (end - start) / 1_000_000);
  }
```

### CODE REVIEW

**Comment by phulin** [comment]
This test is passing locally. Something must be different about file paths on github. I'm not quite sure what the issue is - any ideas?

**Review by midgleyc — CHANGES_REQUESTED** [review]
Some comments requesting clarification; I can help look into the test thing.

**Comment by midgleyc** [comment]
The tests fail for me locally too: the preferences file is always blank.

**Comment by midgleyc** [comment]
Found the issue: the username wants to be lowercase. You're presumably on a case-insensitive file system?

File: src/net/sourceforge/kolmafia/preferences/Preferences.java [hunk 1]
```diff
@@ -1076,38 +1095,70 @@ private static void setObject(
...
-    } else if (Preferences.userPropertiesFile != null) {
-      // We might be changing the structure of the map.
-      // userValues is a synchronized map.
-
-      Preferences.userValues.put(name, object);
-      if (Preferences.getBoolean("saveSettingsOnSet")) {
-        Preferences.saveToFile(Preferences.userPropertiesFile, Preferences.userValues);
-      }
+    boolean saveSettings = Preferences.getBoolean("saveSettingsOnSet");
+    if (name == "saveSettingsOnSet" && (boolean) object) {
```
[the remark this answers is not in the document]
Comment (phulin) [reply src/net/sourceforge/kolmafia/preferences/Preferences.java:1099]: Encoded values are initialized in `putUser` as values are loaded. I'll rename this "reinitialize" to be more descriptive and add a comment.

File: src/net/sourceforge/kolmafia/preferences/Preferences.java [hunk 2]
```diff
@@ -672,10 +675,26 @@ private static String encodeProperty(String name, String value) {
       buffer.append("=");
       Preferences.encodeString(buffer, value);
     }
+    buffer.append(KoLConstants.LINE_BREAK);
 
     return buffer.toString();
   }
 
+  private static void initializeEncodedValues() {
```
[the remark this answers is not in the document]
Comment (phulin) [reply src/net/sourceforge/kolmafia/preferences/Preferences.java:683]: 👍

**Review by phulin — COMMENTED** [review]
Thanks much for the help.

**Review by midgleyc — APPROVED** [review]

**Pull request merged**